# Análise dos dados da Biblioteca Nacional Digital de Portugal

Nesse notebook, vamos limpar e analisar os dados disponibilizados pela BNP em seu sítio de dados abertos. Buscaremos compreender o escopo do acervo digitalizado e disponibilizado e ao mesmo tempo analisar suas características e como impactam na pesquisa e uso do acervo.

## Parte 1: Dados Gerais

In [1]:
# Importar bibliotecas
import pandas as pd
import plotly.express as px

In [2]:
# importar dataset e criar dataframe
file_encoding = 'utf8'        # set file_encoding to the file encoding (utf8, latin1, etc.)
input_fd = open("/home/ebn/Documentos/Github/hemdig-framework/repositorios/BND-PT/bnd-pt.csv", encoding=file_encoding, errors = 'backslashreplace')
df = pd.read_csv(input_fd, sep=',', encoding='utf-8')

Sobre o código anterior, ver a [resposta no StackOverflow](https://stackoverflow.com/a/51763708/12751702)

### Visão geral dos dados

As próximas três células apresentam dados comuns aos datasets de obrras em geral e em domínio público.

Primeiras 10 linhas do *dataframe*:

In [3]:
df.head()

BNP record ID Material type ISBN Legal deposit number Language of Text  \
0            544          Book  NaN                  NaN              por   
1            619          Book  NaN                  NaN              por   
2            879          Book  NaN                  NaN              eng   
3           1825          Book  NaN                  NaN              por   
4           2534          Book  NaN                  NaN              por   

  Language of Original Work  \
0                       NaN   
1                       NaN   
2                       NaN   
3                       NaN   
4                       NaN   

                                               Title  \
0  A Suma Oriental de Tomé Pires e o livro de Fra...   
1                  Timor, do século XVI ao século XX   
2                               Francisco de Holanda   
3  O problema da dupla revisão na Constituição Po...   
4                             Elisa de Sousa Pedroso   

                                            Subtitle Original title Edition  \
0                                                NaN            NaN     NaN   
1  [catálogo da exposição] no ano do IV Centenári...            NaN     NaN   
2  a little known source for the history of forti...            NaN     NaN   
3                                                NaN            NaN     NaN   
4                                                NaN            NaN     NaN   

   ...          Name of Publisher Date of Publication Extent of Item  \
0  ...  por Ordem da Universidade                1978      X, 503 p.   
1  ...                   B. N. L.                1980     66, [1] p.   
2  ...  Fund. Calouste Gulbenkian                1979     p. 163-202   
3  ...                     CODECO                1978     38, [1] p.   
4  ...                      [s.n.                1976           4 p.   

  Dimensions                             Series Volume  \
0      22 cm  Acta Universitatis Conimbrigensis    NaN   
1      29 cm                                NaN    NaN   
2      25 cm                                NaN    NaN   
3      21 cm                                NaN    NaN   
4      25 cm                                NaN    NaN   

       Universal Decimal Classification  \
0  910.4(469+5)"15" ; 821.134.3-992"15"   
1                    011(594.75)"15/19"   
2                                    74   
3                                   342   
4                                    92   

                                             Authors  \
0  Pires, Tomé ; Rodrigues, Francisco ; Cortesão,...   
1            Portugal, Biblioteca Nacional de Lisboa   
2                                         Bury, John   
3                             Canotilho, J. J. Gomes   
4                                                NaN   

                                           Image  \
0  https://purl.pt/37597/service/media/cover/low   
1    https://purl.pt/206/service/media/cover/low   
2  https://purl.pt/36983/service/media/cover/low   
3  https://purl.pt/31079/service/media/cover/low   
4  https://purl.pt/39062/service/media/cover/low   

                                Persistent URL  
0   http://id.bnportugal.gov.pt/bib/catbnp/544  
1   http://id.bnportugal.gov.pt/bib/catbnp/619  
2   http://id.bnportugal.gov.pt/bib/catbnp/879  
3  http://id.bnportugal.gov.pt/bib/catbnp/1825  
4  http://id.bnportugal.gov.pt/bib/catbnp/2534  

[5 rows x 21 columns]

O *dataframe* é composto pelas seguintes colunas:

In [4]:
# mostrar colunas em lista
df.columns.tolist()


['BNP record ID',
 'Material type',
 'ISBN',
 'Legal deposit number',
 'Language of Text',
 'Language of Original Work',
 'Title',
 'Subtitle',
 'Original title',
 'Edition',
 'Place of publicattion',
 'Name of Publisher',
 'Date of Publication',
 'Extent of Item',
 'Dimensions',
 'Series',
 'Volume',
 'Universal Decimal Classification',
 'Authors',
 'Image',
 'Persistent URL']

#### Tipos de Materiais

Primeiro vamos analisar que tipos de materiais compõem o acervo digitalizado e disponibilizado pela BNP.

In [5]:
#listar tipos de materiais
df['Material type'].unique().tolist()

['Book',
 'Periodical',
 'Iconographic material',
 nan,
 'Notated music',
 'Cartographic material',
 'Manuscript',
 'Multimedia',
 'Electronic resource']

Encontramos oito tipos de materiais: Livro, Material Iconográfico, Periódico, Material Cartográfico, Partitura, Manuscrito, Fonte eletrônica, Multimeios.

A contagem dos tipos de materiais no conjunto de dados das obras em geral pode ser vista no *dataframe* e gráfico abaixo:

In [6]:
# criar dataframe com os tipos de materiais e a quantidade de cada um
df_material = df['Material type'].value_counts().rename_axis('Material type').reset_index(name='counts')
df_material

Material type  counts
0  Iconographic material   19478
1                   Book   10029
2             Manuscript    2793
3  Cartographic material    2445
4             Periodical    1819
5          Notated music    1029
6             Multimedia      54
7    Electronic resource      43

In [7]:
# criar gráfico de barras com os dados de tipos de materiais
fig1 = px.bar(df_material, x='Material type', y='counts', title='Tipos de materiais')

In [8]:
fig1.show()

Vamos analisar os tipos de materiais por ano de publicação.

In [9]:
# criar dataframe com os tipos de materiais e o ano de publicação e a quantidade de cada um
df_material_ano = df.groupby(['Material type', 'Date of Publication']).size().reset_index(name='counts')
df_material_ano

Material type Date of Publication  counts
0             Book                           3
1             Book                14         3
2             Book                1440       1
3             Book                1454       1
4             Book                1463       1
...            ...                 ...     ...
2352    Periodical                1987       1
2353    Periodical                1989       1
2354    Periodical                1990       1
2355    Periodical                1997       1
2356    Periodical                19_        1

[2357 rows x 3 columns]

In [10]:
# criar gráfico de barras com os dados de tipos de materiais e ano de publicação
fig2 = px.bar(df_material_ano, x='Material type', y='counts', color='Date of Publication', title='Tipos de materiais e ano de publicação')
fig2.show()

Vamos analisar o volume de itens por ano de publicação, independentemente do tipo de material.

In [11]:
# Criar dataframe com a quantidade de publicações por ano
df_ano = df['Date of Publication'].value_counts().rename_axis('Date of Publication').reset_index(name='counts')
# sort dataframe by year
df_ano = df_ano.sort_values(by=['Date of Publication'])
df_ano

Date of Publication  counts
548                           6
658                1010       1
637                1075       1
541                1101       6
634                1111       1
..                  ...     ...
275                2018      39
498                2019      13
500                2020      13
553                2021       6
686                2022       1

[687 rows x 2 columns]

In [12]:
# Criar scatter plot com os dados de datas de publicação com as datas de publicação no eixo Y
fig3 = px.scatter(df_ano, x='Date of Publication', y='counts', title='Quantidades de publicação por ano - Todos os materiais')
fig3.show()

### Análise do acervo de Periódicos

Vamos analisar com mais detalhes o acervo digitalizado de periódicos.

In [13]:
#criar df apenas com coluna "Periodical"
df_periodical = df[df['Material type'] == 'Periodical']

In [14]:
df_periodical.head()


BNP record ID Material type ISBN Legal deposit number Language of Text  \
11            7101    Periodical  NaN                  NaN              por   
370          33485    Periodical  NaN                  NaN              por   
436          37772    Periodical  NaN                  NaN              por   
437          40112    Periodical  NaN                  NaN              por   
745          57737    Periodical  NaN                  NaN              por   

    Language of Original Work            Title  \
11                        NaN       Telescópio   
370                       NaN         Sudoeste   
436                       NaN    Obras Avulsas   
437                       NaN          Porbase   
745                       NaN  O Pinto Calçudo   

                                  Subtitle Original title Edition  ...  \
11                                     NaN            NaN     NaN  ...   
370           cadernos de Almada Negreiros            NaN     NaN  ...   
436                                    NaN            NaN   2ª ed  ...   
437  Base Nacional de Dados Bibliográficos            NaN     NaN  ...   
745                                    NaN            NaN     NaN  ...   

                                     Name of Publisher Date of Publication  \
11                                                CEPA                1978   
370                                   Contexto Editora                1982   
436                                           E.N.S.P.                198    
437                                Biblioteca Nacional                1987   
745  Alunos da Escola Masculina de Vila Real de San...                1958   

    Extent of Item Dimensions    Series Volume  \
11             NaN      35 cm       NaN    NaN   
370            NaN      25 cm  Contexto      5   
436            NaN      24 cm       NaN    NaN   
437            NaN      30 cm       NaN    NaN   
745            NaN      27 cm       NaN    NaN   

          Universal Decimal Classification  \
11                                  52(05)   
370  869.0(05) ; 7.036(05) ; 82.015.19(05)   
436                                    NaN   
437                   002.6:681.3(469)(05)   
745            372.22(469.601.26 VRSA)(05)   

                                               Authors  \
11                  Magalhães, Alberto ; CEPA, Alberto   
370                                     Martins, Dario   
436  Universidade Nova de Lisboa, Escola Nacional d...   
437                      Portugal, Biblioteca Nacional   
745  Borges, José Maria S. Baptista ; Marques, Carl...   

                                             Image  \
11   https://purl.pt/34992/service/media/cover/low   
370  https://purl.pt/33835/service/media/cover/low   
436  https://purl.pt/39633/service/media/cover/low   
437  https://purl.pt/23861/service/media/cover/low   
745  https://purl.pt/36015/service/media/cover/low   

                                   Persistent URL  
11    http://id.bnportugal.gov.pt/bib/catbnp/7101  
370  http://id.bnportugal.gov.pt/bib/catbnp/33485  
436  http://id.bnportugal.gov.pt/bib/catbnp/37772  
437  http://id.bnportugal.gov.pt/bib/catbnp/40112  
745  http://id.bnportugal.gov.pt/bib/catbnp/57737  

[5 rows x 21 columns]

#### Idiomas

In [15]:
# contar idiomas dos periódicos
df_periodical['Language of Text'].value_counts()

por       1798
fre          7
eng          7
spa          3
por          1
mul          1
chi          1
ger          1
Name: Language of Text, dtype: int64

In [16]:
# criar gráfico de barras com os dados de idiomas dos periódicos
fig4 = px.bar(df_periodical, x='Language of Text', title='Idiomas dos periódicos')
fig4.show()

Praticamente a totalidade dos itens são em português, com uma quantidade estatisticamente insignificante de itens em outros idiomas.

#### Data de Publicação

Vamos analisar a distribuição dos itens por ano de publicação e buscar compreender quais períodos históricos estão mais bem representados no acervo.

In [17]:
# criar dataframe com as datas de publicação e a quantidade de cada uma organizadas cronologicamente
df_date = df_periodical['Date of Publication'].value_counts().rename_axis('Date of Publication').reset_index(name='counts')
# organizaro o dataframe pelos valores na coluna "Date of Publication"
df_date = df_date.sort_values(by=['Date of Publication'])
df_date

Date of Publication  counts
129                1641       1
153                1730       1
139                1743       1
149                1752       1
141                1779       1
..                  ...     ...
158                1987       1
131                1989       1
145                1990       1
135                1997       1
170                19_        1

[179 rows x 2 columns]

In [18]:
# Criar scatter plot com os dados de datas de publicação com as datas de publicação no eixo Y
fig5 = px.scatter(df_date, x='Date of Publication', y='counts', title='Quantidade de publicação dos periódicos por ano')
fig5.show()

Podemos perceber que o escopo temporal do acervo se estende desde meados do século XVII até o último quartel do século XX. A primeira publicação data de 1641 e a última de 1997.

Entretanto, os anos de publicação com mais de 10 itens se concentram entre 1872 e 1932.

In [19]:
# remove all special characters in column "Date of Publication"
df_periodical['Date of Publication'] = df_periodical['Date of Publication'].str.replace(r'\W', '')
df_periodical['Date of Publication'] = df_periodical['Date of Publication'].str.replace('_', '')


/tmp/ipykernel_158928/1120145348.py:2: FutureWarning:

The default value of regex will change from True to False in a future version.

/tmp/ipykernel_158928/1120145348.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/1120145348.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [20]:
# transformar a coluna "Date of Publication" em inteiro
df_periodical['Date of Publication'] = df_periodical['Date of Publication'].astype(int)

/tmp/ipykernel_158928/2299171753.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [21]:
# count unique values in column date of publication and sum them
df_periodical['Date of Publication'].nunique()

177

In [22]:
# count values in column date of publication and sum them
df_periodical['Date of Publication'].count()

1819

In [23]:
# count values between 1872 and 1920 and sum them
df_periodical[(df_periodical['Date of Publication'] >= 1872) & (df_periodical['Date of Publication'] <= 1932)]['Date of Publication'].value_counts().sum()

1475

Nos 60 anos entre 1872 e 1932, encontramos 1475 itens, o que dá uma média de 24,6 itens por anos e corresponde a 81,1%. Nos 117 anos restantes, encontramos 344 itens, o que dá uma média de 3,2 itens por ano e corresponde a 18,9%.

In [24]:
# criar dataframe com os dados
df_periodical_1872_1932 = pd.DataFrame({'1872-1932': [1475], 'Demais Anos': [344]})
#criar gráfico de barras com os dados
fig6 = px.bar(df_periodical_1872_1932, title='Periódicos por ano de publicação')
fig6.show()

In [25]:
# criar dataframe com as datas de publicação que contam apenas 10 ou mais publicações
df_date_10 = df_date[df_date['counts'] >= 10]
df_date_10

Date of Publication  counts
55                1809      10
41                1826      20
53                1836      11
56                1839      10
58                1872      10
43                1875      18
51                1876      12
54                1877      11
39                1878      20
42                1879      19
22                1880      29
21                1881      29
16                1882      31
20                1883      29
30                1884      24
4                 1885      39
24                1886      29
10                1887      34
8                 1888      38
9                 1889      37
23                1890      29
2                 1891      40
26                1892      26
5                 1893      39
1                 1894      41
6                 1895      38
0                 1896      45
3                 1897      40
17                1898      30
31                1899      24
44                1900      18
57                1902      10
46                1903      14
50                1905      13
49                1906      13
37                1907      22
35                1908      23
15                1909      32
7                 1910      38
14                1911      32
12                1912      34
13                1913      33
11                1914      34
28                1915      25
25                1916      26
32                1917      24
45                1918      18
19                1919      30
38                1920      21
34                1921      24
27                1922      26
33                1923      24
36                1924      23
18                1925      30
29                1926      25
47                1927      14
40                1929      20
48                1930      14
52                1932      11

In [26]:
# transforme a coluna "Date of Publication" em inteiro usando .loc
df_date_10.loc[:, 'Date of Publication'] = df_date_10.loc[:, 'Date of Publication'].astype(int)
df_date_10

/tmp/ipykernel_158928/3507128909.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/3507128909.py:2: FutureWarning:

In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`



Date of Publication  counts
55                 1809      10
41                 1826      20
53                 1836      11
56                 1839      10
58                 1872      10
43                 1875      18
51                 1876      12
54                 1877      11
39                 1878      20
42                 1879      19
22                 1880      29
21                 1881      29
16                 1882      31
20                 1883      29
30                 1884      24
4                  1885      39
24                 1886      29
10                 1887      34
8                  1888      38
9                  1889      37
23                 1890      29
2                  1891      40
26                 1892      26
5                  1893      39
1                  1894      41
6                  1895      38
0                  1896      45
3                  1897      40
17                 1898      30
31                 1899      24
44                 1900      18
57                 1902      10
46                 1903      14
50                 1905      13
49                 1906      13
37                 1907      22
35                 1908      23
15                 1909      32
7                  1910      38
14                 1911      32
12                 1912      34
13                 1913      33
11                 1914      34
28                 1915      25
25                 1916      26
32                 1917      24
45                 1918      18
19                 1919      30
38                 1920      21
34                 1921      24
27                 1922      26
33                 1923      24
36                 1924      23
18                 1925      30
29                 1926      25
47                 1927      14
40                 1929      20
48                 1930      14
52                 1932      11

In [27]:
# criar gráfico de barras com os dados de datas de publicação que contam apenas 10 ou mais publicações, organizadas cronologicamente
fig7 = px.bar(df_date_10, x='Date of Publication', y='counts', title='Quantidade de publicação dos periódicos - 10 ou mais por ano')
fig7.show()

Todas os anos com mais de 10 publicações estão concentrados no século XIX e primeira metade do século XX.

#### Editoras e editores

In [28]:
# na coluna name of publisher, substituir os valores ""[s.n.]", "s.n.]" e "[s.n." por "Sem editora"
df_periodical['Name of Publisher'] = df_periodical['Name of Publisher'].replace(['[s.n.]', 's.n.]', '[s.n.', 's.n.'], 'Sem editora')
# unir os valores "na Impressão Regia" e "na Impressão Régia"
df_periodical['Name of Publisher'] = df_periodical['Name of Publisher'].replace(['na Impressão Regia', 'na Impressão Régia'], 'na Impressão Régia')
# unir os valores "Imprensa Nacional" e "na Imprensa Nacional"
df_periodical['Name of Publisher'] = df_periodical['Name of Publisher'].replace(['Imprensa Nacional', 'na Imprensa Nacional'], 'na Imprensa Nacional')


/tmp/ipykernel_158928/2599309952.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/2599309952.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/2599309952.py:6: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [29]:
# criar dataframe com as editoras e a quantidade de cada uma
df_publisher = df_periodical['Name of Publisher'].value_counts().rename_axis('Name of Publisher').reset_index(name='counts')
df_publisher

Name of Publisher  counts
0              Sem editora     190
1         Paulo da Fonseca      17
2     na Imprensa Nacional      16
3        J. Garcia de Lima      13
4       na Impressão Régia      12
...                    ...     ...
1397   na Typ. do Panorama       1
1398          J. S. Vieira       1
1399                E.E.A.       1
1400             A. Birros       1
1401  Tip. Minerva Indiana       1

[1402 rows x 2 columns]

In [30]:
# criar um gráfico de barras com os 50 primeiros valores de editoras
fig8 = px.bar(df_publisher.head(50), x='Name of Publisher', y='counts', title='Editoras dos periódicos')
fig8.show()

In [31]:
# criar um gráfico de barras com as primeiras 20 editoras, excluindo a primeira.
fig9 = px.bar(df_publisher[1:31], x='Name of Publisher', y='counts', title='Editoras dos periódicos (sem itens "Sem editora")')
fig9.show()

#### Contar palavras nos títulos e subtítulos

In [32]:
# criar dataframe com a quantidade de títulos
df_title = df_periodical['Title'].value_counts().rename_axis('Title').reset_index(name='counts')
df_title

Title  counts
0                A Verdade      12
1                 O Futuro      11
2              O Academico       9
3               A Alvorada       7
4                O Liberal       6
...                    ...     ...
1540         O Legitimista       1
1541  O Jornal do Operario       1
1542       O Contemporaneo       1
1543   Jornal dos Romances       1
1544           A Provincia       1

[1545 rows x 2 columns]

In [33]:
# ler a coluna título e contar palavras únicas e a quantidade de vezes que aparecem
#mostrar apenas palavras mais longas que 3 caracteres
# mostrar apenas as 100 palavras mais frequentes
df_unique_title = df_title['Title'].str.split(expand=True).stack().value_counts().loc[lambda x : x.index.str.len() > 3].head(100)
df_unique_title = df_unique_title.rename_axis('Title').reset_index(name='counts')
df_unique_title

Title  counts
0      Jornal      52
1     Revista      42
2     Correio      35
3    Portugal      31
4      Gazeta      29
..        ...     ...
95       João       4
96      Echos       4
97     Brazil       4
98    Justiça       4
99  Exposição       4

[100 rows x 2 columns]

In [34]:
# gráfico de barras com as 50 palavras mais frequentes
fig10 = px.bar(df_unique_title.head(50), x='Title', y='counts', title='Palavras mais repetidas nos títulos dos periódicos (com mais de três caracteres)')
fig10.show()

In [35]:
# criar dataframe com a quantidade de títulos
df_subtitle = df_periodical['Subtitle'].value_counts().rename_axis('Subtitle').reset_index(name='counts')
df_subtitle

Subtitle  counts
0                                semanario independente      12
1                                 semanario republicano      11
2                                         folha semanal      11
3                                 semanário republicano       7
4                                 semanario humoristico       5
...                                                 ...     ...
1028                                    folha imparcial       1
1029  número commemorativo offerecido por J. C. Viei...       1
1030                      jornal d´annuncios illustrado       1
1031               quinzenario litterario e humoristico       1
1032                                   mensário de arte       1

[1033 rows x 2 columns]

In [36]:
# ler a coluna título e contar palavras únicas e a quantidade de vezes que aparecem
#mostrar apenas palavras mais longas que 3 caracteres
# mostrar apenas as 100 palavras mais frequentes
df_unique_subtitle = df_subtitle['Subtitle'].str.split(expand=True).stack().value_counts().loc[lambda x : x.index.str.len() > 3].head(100)
df_unique_subtitle = df_unique_subtitle.rename_axis('Subtitle').reset_index(name='counts')
df_unique_subtitle

Subtitle  counts
0      semanario     136
1         jornal     118
2        revista     112
3          orgão     106
4          folha      81
..           ...     ...
95       popular       7
96         arte,       7
97   democratico       7
98  litteratura,       7
99     Sociedade       6

[100 rows x 2 columns]

In [37]:
# gráfico de barras com as 50 palavras mais frequentes
fig11 = px.bar(df_unique_subtitle.head(50), x='Subtitle', y='counts', title='Palavras mais repetidas nos subtítulos dos periódicos (com mais de três caracteres)')
fig11.show()

:warning: Os dados acima precisam ser limpos e analisados com mais calma. Talvez seja mais vantajoso submetê-los a uma análise de tópicos ou NER com o spaCy.

### Considerações gerais

Acervo de periódicos é sobretudo em português, e seu recorte temporal é vasto, desde o século XVII até o século XX. Contudo, ao analisarmos os dados percebemos que a maior parte dos itens se concentra entre 1872 e 1920: atingindo mais de 75% de todos os itens digitalizados.

## Parte 2: Dados de de acervo em domínio público

Nessa segunda parte, repetiremos a análise anterior, mas considerando apenas os itens em domínio público.

In [38]:
# importar dataset e criar dataframe
file_encoding = 'utf8'        # set file_encoding to the file encoding (utf8, latin1, etc.)
input_fd = open("/home/ebn/Documentos/Github/hemdig-framework/repositorios/BND-PT/bnd-pt_livre.csv", encoding=file_encoding, errors = 'backslashreplace')
df_livre = pd.read_csv(input_fd, sep=',', encoding='utf-8')

In [39]:
# criar dataframe com os tipos de materiais e a quantidade de cada um
df_material_livre = df_livre['Material type'].value_counts().rename_axis('Material type').reset_index(name='counts')
df_material_livre

Material type  counts
0                   Book    9008
1  Iconographic material    7953
2             Manuscript    2781
3  Cartographic material    2283
4             Periodical    1359
5          Notated music     948
6             Multimedia      52
7    Electronic resource      35

In [40]:
# criar gráfico de barras com os dados de tipos de materiais
fig12 = px.bar(df_material_livre, x='Material type', y='counts', title='Tipos de materiais em domínio público.')
fig12.show()

In [41]:
# criar dataframe com os tipos de materiais e o ano de publicação e a quantidade de cada um
df_material_ano_livre = df_livre.groupby(['Material type', 'Date of Publication']).size().reset_index(name='counts')
df_material_ano_livre

Material type Date of Publication  counts
0             Book                           3
1             Book                14         3
2             Book                1440       1
3             Book                1454       1
4             Book                1463       1
...            ...                 ...     ...
2226    Periodical                1964       1
2227    Periodical                1975       1
2228    Periodical                1987       1
2229    Periodical                1989       1
2230    Periodical                19_        1

[2231 rows x 3 columns]

In [42]:
# criar gráfico de barras com os dados de tipos de materiais e ano de publicação
fig13 = px.bar(df_material_ano_livre, x='Material type', y='counts', color='Date of Publication', title='Tipos de materiais e ano de publicação em domínio público.')
fig13.show()

In [43]:
# Criar dataframe com a quantidade de publicações por ano
df_ano_livre = df_livre['Date of Publication'].value_counts().rename_axis('Date of Publication').reset_index(name='counts')
# sort dataframe by year
df_ano_livre = df_ano_livre.sort_values(by=['Date of Publication'])
df_ano_livre

Date of Publication  counts
547                           5
659                1010       1
667                1075       1
526                1101       6
624                1111       1
..                  ...     ...
357                2015      23
194                2016      43
144                2017      52
321                2018      27
537                2019       5

[679 rows x 2 columns]

In [44]:
# Criar scatter plot com os dados de datas de publicação com as datas de publicação no eixo Y
fig14 = px.scatter(df_ano_livre, x='Date of Publication', y='counts', title='Quantidade de publicação em domínio público por ano.')
fig14.show()

In [45]:
#criar df apenas com coluna "Periodical"
df_periodical_livre = df_livre[df_livre['Material type'] == 'Periodical']

In [46]:
df_periodical_livre.head()

BNP record ID Material type ISBN Legal deposit number Language of Text  \
52           40112    Periodical  NaN                  NaN              por   
147          57737    Periodical  NaN                  NaN              por   
159          63466    Periodical  NaN                  NaN              por   
230          90413    Periodical  NaN                  NaN              por   
231          90414    Periodical  NaN                  NaN              por   

    Language of Original Work            Title  \
52                        NaN          Porbase   
147                       NaN  O Pinto Calçudo   
159                       NaN     O João Ratão   
230                       NaN    Alma Nacional   
231                       NaN          A Águia   

                                              Subtitle Original title Edition  \
52               Base Nacional de Dados Bibliográficos            NaN     NaN   
147                                                NaN            NaN     NaN   
159                                                NaN            NaN     NaN   
230                                                NaN            NaN     NaN   
231  revista quinzenal ilustrada de literatura e cr...            NaN     NaN   

     ...                                  Name of Publisher  \
52   ...                                Biblioteca Nacional   
147  ...  Alunos da Escola Masculina de Vila Real de San...   
159  ...              Caixa Escolar Nossa Senhora de Fátima   
230  ...                                      A. J. Almeida   
231  ...                                     Tércio Miranda   

    Date of Publication Extent of Item Dimensions Series Volume  \
52                 1987            NaN      30 cm    NaN    NaN   
147                1958            NaN      27 cm    NaN    NaN   
159                1961            NaN      20 cm    NaN    NaN   
230                1910            NaN      26 cm    NaN    NaN   
231                1910            NaN      31 cm    NaN    NaN   

    Universal Decimal Classification  \
52              002.6:681.3(469)(05)   
147      372.22(469.601.26 VRSA)(05)   
159                 372(469.601)(05)   
230                              NaN   
231                              070   

                                               Authors  \
52                       Portugal, Biblioteca Nacional   
147  Borges, José Maria S. Baptista ; Marques, Carl...   
159  Caixa Escolar Nossa Senhora de Fátima ; Escola...   
230                           Almeida, António José de   
231  Pinto, Álvaro ; Pascoais, Teixeira de ; Carnei...   

                                             Image  \
52   https://purl.pt/23861/service/media/cover/low   
147  https://purl.pt/36015/service/media/cover/low   
159  https://purl.pt/36019/service/media/cover/low   
230  https://purl.pt/16522/service/media/cover/low   
231  https://purl.pt/12152/service/media/cover/low   

                                   Persistent URL  
52   http://id.bnportugal.gov.pt/bib/catbnp/40112  
147  http://id.bnportugal.gov.pt/bib/catbnp/57737  
159  http://id.bnportugal.gov.pt/bib/catbnp/63466  
230  http://id.bnportugal.gov.pt/bib/catbnp/90413  
231  http://id.bnportugal.gov.pt/bib/catbnp/90414  

[5 rows x 21 columns]

In [47]:
# contar idiomas dos periódicos
df_periodical_livre['Language of Text'].value_counts()

por    1343
eng       6
fre       5
spa       2
mul       1
chi       1
ger       1
Name: Language of Text, dtype: int64

In [48]:
fig15 = px.bar(df_periodical, x='Language of Text', title='Idiomas dos periódicos em domínio público.')
fig15.show()

In [49]:
# criar dataframe com as datas de publicação e a quantidade de cada uma organizadas cronologicamente
df_date_livre = df_periodical_livre['Date of Publication'].value_counts().rename_axis('Date of Publication').reset_index(name='counts')
# organizaro o dataframe pelos valores na coluna "Date of Publication"
df_date_livre = df_date_livre.sort_values(by=['Date of Publication'])
df_date_livre

Date of Publication  counts
123                1641       1
132                1730       1
105                1743       1
127                1752       1
110                1779       1
..                  ...     ...
116                1964       1
118                1975       1
121                1987       1
122                1989       1
141                19_        1

[154 rows x 2 columns]

In [50]:
# Criar scatter plot com os dados de datas de publicação com as datas de publicação no eixo Y
fig16 = px.scatter(df_date_livre, x='Date of Publication', y='counts', title='Quantidade de publicação dos periódicos em domínio público por ano.')
fig16.show()

In [51]:
# remove all special characters in column "Date of Publication"
df_periodical_livre['Date of Publication'] = df_periodical_livre['Date of Publication'].str.replace(r'\W', '')
df_periodical_livre['Date of Publication'] = df_periodical_livre['Date of Publication'].str.replace('_', '')

/tmp/ipykernel_158928/4270708723.py:2: FutureWarning:

The default value of regex will change from True to False in a future version.

/tmp/ipykernel_158928/4270708723.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/4270708723.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [52]:
# transformar a coluna "Date of Publication" em inteiro
df_periodical_livre['Date of Publication'] = df_periodical_livre['Date of Publication'].astype(int)

/tmp/ipykernel_158928/1228984385.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [53]:
# count unique values in column date of publication and sum them
df_periodical_livre['Date of Publication'].nunique()

152

In [54]:
# count values in column date of publication and sum them
df_periodical_livre['Date of Publication'].count()

1359

In [55]:
# count values between 1872 and 1920 and sum them
df_periodical_livre[(df_periodical_livre['Date of Publication'] >= 1872) & (df_periodical_livre['Date of Publication'] <= 1920)]['Date of Publication'].value_counts().sum()

1024

In [56]:
# criar dataframe com os dados
df_periodical_1872_1920 = pd.DataFrame({'1872-1920': [1024], 'Demais Anos': [335]})
#criar gráfico de barras com os dados
fig17 = px.bar(df_periodical_1872_1920, title='Periódicos em domínio público por ano de publicação.')
fig17.show()

In [57]:
# criar dataframe com as datas de publicação que contam apenas 10 ou mais publicações
df_date_10_livre = df_date_livre[df_date_livre['counts'] >= 10]
df_date_10_livre

Date of Publication  counts
42                1809      10
22                1826      20
36                1836      11
41                1839      10
40                1872      10
27                1875      18
34                1876      12
38                1877      11
23                1878      20
26                1879      18
16                1880      28
11                1881      29
10                1882      31
14                1883      29
19                1884      24
4                 1885      39
12                1886      29
9                 1887      34
5                 1888      38
8                 1889      36
13                1890      29
2                 1891      40
17                1892      26
6                 1893      38
0                 1894      41
7                 1895      37
3                 1896      40
1                 1897      40
15                1898      28
21                1899      23
30                1900      14
37                1903      11
25                1907      18
29                1908      15
18                1909      26
20                1910      23
24                1911      19
35                1912      12
28                1913      17
32                1914      13
31                1917      14
39                1918      11
43                1919      10
33                1920      13

In [58]:
# transforme a coluna "Date of Publication" em inteiro usando .loc
df_date_10_livre.loc[:, 'Date of Publication'] = df_date_10_livre.loc[:, 'Date of Publication'].astype(int)
df_date_10_livre

/tmp/ipykernel_158928/2109869929.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/2109869929.py:2: FutureWarning:

In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`



Date of Publication  counts
42                 1809      10
22                 1826      20
36                 1836      11
41                 1839      10
40                 1872      10
27                 1875      18
34                 1876      12
38                 1877      11
23                 1878      20
26                 1879      18
16                 1880      28
11                 1881      29
10                 1882      31
14                 1883      29
19                 1884      24
4                  1885      39
12                 1886      29
9                  1887      34
5                  1888      38
8                  1889      36
13                 1890      29
2                  1891      40
17                 1892      26
6                  1893      38
0                  1894      41
7                  1895      37
3                  1896      40
1                  1897      40
15                 1898      28
21                 1899      23
30                 1900      14
37                 1903      11
25                 1907      18
29                 1908      15
18                 1909      26
20                 1910      23
24                 1911      19
35                 1912      12
28                 1913      17
32                 1914      13
31                 1917      14
39                 1918      11
43                 1919      10
33                 1920      13

In [59]:
# criar gráfico de barras com os dados de datas de publicação que contam apenas 10 ou mais publicações, organizadas cronologicamente
fig18 = px.bar(df_date_10_livre, x='Date of Publication', y='counts', title='Quantidade de publicação dos periódicos em domínio público - 10 ou mais por ano')
fig18.show()

In [60]:
# na coluna name of publisher, substituir os valores ""[s.n.]", "s.n.]" e "[s.n." por "Sem editora"
df_periodical_livre['Name of Publisher'] = df_periodical_livre['Name of Publisher'].replace(['[s.n.]', 's.n.]', '[s.n.', 's.n.'], 'Sem editora')
# unir os valores "na Impressão Regia" e "na Impressão Régia"
df_periodical_livre['Name of Publisher'] = df_periodical_livre['Name of Publisher'].replace(['na Impressão Regia', 'na Impressão Régia'], 'na Impressão Régia')
# unir os valores "Imprensa Nacional" e "na Imprensa Nacional"
df_periodical_livre['Name of Publisher'] = df_periodical_livre['Name of Publisher'].replace(['Imprensa Nacional', 'na Imprensa Nacional'], 'na Imprensa Nacional')

/tmp/ipykernel_158928/2386316166.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/2386316166.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipykernel_158928/2386316166.py:6: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [61]:
# criar dataframe com as editoras e a quantidade de cada uma
df_publisher_livre = df_periodical_livre['Name of Publisher'].value_counts().rename_axis('Name of Publisher').reset_index(name='counts')
df_publisher_livre

Name of Publisher  counts
0                    Sem editora     171
1               Paulo da Fonseca      17
2           na Imprensa Nacional      16
3              J. Garcia de Lima      13
4             na Impressão Régia      12
..                           ...     ...
986                  Tasso Lopes       1
987  Typographia do Lacobrigense       1
988       Typ. Franco-Portugueza       1
989      P. A. Marchal e Milcent       1
990         Tip. Minerva Indiana       1

[991 rows x 2 columns]

In [62]:
# criar um gráfico de barras com os 50 primeiros valores de editoras
fig19 = px.bar(df_publisher_livre.head(50), x='Name of Publisher', y='counts', title='Editoras dos periódicos em domínio público.')
fig19.show()

In [63]:
# criar um gráfico de barras com as primeiras 20 editoras, excluindo a primeira.
fig20 = px.bar(df_publisher_livre[1:31], x='Name of Publisher', y='counts', title='Editoras dos periódicos em domínio público (sem itens "Sem editora")')
fig20.show()

In [64]:
# criar dataframe com a quantidade de títulos
df_title_livre = df_periodical_livre['Title'].value_counts().rename_axis('Title').reset_index(name='counts')
df_title_livre

Title  counts
0                O Academico       8
1                  A Verdade       7
2                 A Alvorada       6
3                A Liberdade       6
4                  O Liberal       5
...                      ...     ...
1189               O Japonez       1
1190                O Jornal       1
1191      Jornal da Bairrada       1
1192  Gazeta dos Telegraphos       1
1193             A Provincia       1

[1194 rows x 2 columns]

In [65]:
# ler a coluna título e contar palavras únicas e a quantidade de vezes que aparecem
#mostrar apenas palavras mais longas que 3 caracteres
# mostrar apenas as 100 palavras mais frequentes
df_unique_title_livre = df_title_livre['Title'].str.split(expand=True).stack().value_counts().loc[lambda x : x.index.str.len() > 3].head(100)
df_unique_title_livre = df_unique_title_livre.rename_axis('Title').reset_index(name='counts')
df_unique_title_livre

Title  counts
0        Jornal      45
1       Revista      34
2        Gazeta      26
3      Portugal      26
4       Correio      23
..          ...     ...
95        Novas       3
96    Federação       3
97       Seculo       3
98  Agricultura       3
99        Lucta       3

[100 rows x 2 columns]

In [66]:
# gráfico de barras com as 50 palavras mais frequentes
fig21 = px.bar(df_unique_title_livre.head(50), x='Title', y='counts', title='Palavras mais frequentes nos títulos dos periódicos em domínio público.')
fig21.show()

In [67]:
# criar dataframe com a quantidade de títulos
df_subtitle_livre = df_periodical_livre['Subtitle'].value_counts().rename_axis('Subtitle').reset_index(name='counts')
df_subtitle_livre

Subtitle  counts
0                               semanario independente      10
1                                semanario republicano      10
2                                        folha semanal       8
3           semanario politico, litterario e noticioso       5
4                                semanario humoristico       5
..                                                 ...     ...
740                           folha assim e assado ...       1
741                                 folha hebdomadaria       1
742                     publicação semanal, illustrada       1
743  jornal illustrado, theatros musica e bellas-artes       1
744                   [publicação mensal e patriotica]       1

[745 rows x 2 columns]

In [68]:
# ler a coluna título e contar palavras únicas e a quantidade de vezes que aparecem
#mostrar apenas palavras mais longas que 3 caracteres
# mostrar apenas as 100 palavras mais frequentes
df_unique_subtitle_livre = df_subtitle_livre['Subtitle'].str.split(expand=True).stack().value_counts().loc[lambda x : x.index.str.len() > 3].head(100)
df_unique_subtitle_livre = df_unique_subtitle_livre.rename_axis('Subtitle').reset_index(name='counts')
df_unique_subtitle_livre

Subtitle  counts
0     semanario     107
1        jornal     101
2       revista      82
3         folha      73
4    litterario      66
..          ...     ...
95    portuense       4
96   noticiosa,       4
97  caricaturas       4
98     chronica       4
99      lettras       4

[100 rows x 2 columns]

In [69]:
# gráfico de barras com as 50 palavras mais frequentes
fig22 = px.bar(df_unique_subtitle_livre.head(50), x='Subtitle', y='counts', title='Palavras mais frequentes nos subtítulos dos periódicos em domínio público.')
fig22.show()

## Comparação entre acervo geral e domínio público

Vamos comparar os dois conjuntos de dados e verificar se há diferenças significativas entre eles. Nesse momento, vamos comparar apenas os gráficos.

### Comparação entre os tipos de materiais


In [70]:
# mostrar fig1 e fig12 
fig1.show()
fig12.show()


### Tipos de materiais e ano de publicação

In [71]:
# mostrar fig2 e fig13
fig2.show()
fig13.show()

### Quantidades de publicação por ano - Todos os materiais

In [72]:
# mostrar fig3 e fig14
fig3.show()
fig14.show()

### Periódicos - Comparação entre acervo geral e domínio público

#### Idiomas

In [73]:
# mostrar fig4 e fig15
fig4.show()
fig15.show()

#### Quantidade de publicação dos periódicos por ano

In [74]:
# mostrar fig5 e fig16
fig5.show()
fig16.show()

#### Período de concentração de publicações no acervo

In [75]:
# mostrar fig6 e fig17
fig6.show()
fig17.show()

#### Quantidade de publicação dos periódicos - 10 ou mais por ano

In [76]:
# mostrar fig7 e fig18
fig7.show()
fig18.show()

#### Editoras dos periódicos

In [77]:
# mostrar fig8 e fig19
fig8.show()
fig19.show()

##### Editoras dos periódicos (sem itens "Sem editora")

In [78]:
# mostrar fig9 e fig20
fig9.show()
fig20.show()

#### Palavras mais repetidas nos títulos e subtítulos dos periódicos

##### Títulos dos periódicos (com mais de três caracteres)

In [79]:
# mostrar fig10 e fig21
fig10.show()
fig21.show()

##### Subtítulos dos periódicos (com mais de três caracteres)

In [80]:
# mostrar fig11 e fig22
fig11.show()
fig22.show()